### PR TITLE
[WIP] first version of fatal errors

### PR DESCRIPTION
**Pull request by jreidinger**
plan is to also add all current checks in bootloader itself.
screenshot from manual testing:

snimek obrazovky_2018-02-15_09-44-19

and warnings still works:

snimek obrazovky_2018-02-15_09-48-26

### CODE REVIEW

File: src/lib/y2storage/boot_requirements_strategies/prep.rb [hunk 1]
```diff
@@ -60,6 +56,9 @@ def errors
     protected
 
       def boot_partition_needed?
+        # FIXME: I (JR) think this is needed only for powernv where firmware have
+        #        to load grub2 configuration and kernel itself. For prep case
+        #        grub2 loads it itself.
```
Comment (joseivanlopez) [inline src/lib/y2storage/boot_requirements_strategies/prep.rb]: Yes. I am fixing this here https://github.com/yast/yast-storage-ng/pull/520

File: src/lib/y2storage/boot_requirements_strategies/zipl.rb [hunk 2]
```diff
@@ -48,6 +46,17 @@ def errors
         errors
       end
 
+      def fatal_errors
+        res = super
+        return res unless res.empty?
+
+        if missing_partition_for?(zipl_volume)
```
Comment (joseivanlopez) [inline src/lib/y2storage/boot_requirements_strategies/zipl.rb]: Why to check with a volume and report the error with another?

Comment (jreidinger) [reply src/lib/y2storage/boot_requirements_strategies/zipl.rb]: @joseivanlopez bug...btw in bootloader check is that /boot/zip *path* is on ext4. So we are fine if /boot is ext4 and no separate /boot/zipl needed or even / as ext4. how I can express it with volume?

File: src/lib/y2storage/boot_requirements_strategies/base.rb [hunk 3]
```diff
@@ -77,9 +77,29 @@ def needed_partitions(target)
       #
       # @return [Array<SetupError>]
       def errors
+        raise Error, "There are fatal errors, so cannot detect soft errors." unless fatal_errors.empty?
```
Comment (joseivanlopez) [inline src/lib/y2storage/boot_requirements_strategies/base.rb]: Raise an exception here looks me a little bit strong.

**Comment by joseivanlopez** [comment]
A general comment. In some cases we could have both: "fatal errors" and "errors". For example, root is missing (fatal error) and also /boot/efi is missing (error). We could show both errors at the same time, and so the use knows what he needs to fix his configuration. With current version, we show first only fatal errors, and then errors.

Moreover, what about renaming it as "errors" and "warnings"?

I would show all of them together (what makes sense) in the same popup, and I will activate the proper buttons depending on if there are only warnings or not.

**Comment by jreidinger** [comment]
@joseivanlopez I am considering that having together warning and errors and it have one problem that it confuse users. We show popup where is:

1 fatal entry
X boot entries ( e.g. when / missing or partition table I expect at least 5 another warnings )
Y product entries ( with missing / I expect another bunch of warnings )
and there is only 1 fatal.

So I decide to show when we have blocker only blocker to really focus on it. and when we have set of warnings, show all of them, as user have to agree with all of them.

Renaming sounds reasonable for me.

File: src/lib/y2storage/setup_checker.rb [hunk 4]
```diff
@@ -55,43 +55,43 @@ def initialize(devicegraph)
...
-      boot_errors.empty?
+    # @return [Array<SetupError>]
+    def errors
+      boot_requirements_checker.errors
     end
 
-    # All storage errors detected in the setup, for example, when a /boot/efi partition
-    # is missing in a UEFI system, or a valid partition for root is missing.
+    # All storage warnings detected in the setup, for example, when a /boot/efi partition
+    # is missing in a UEFI system, or missing separate /boot.
```
Comment (joseivanlopez) [inline src/lib/y2storage/setup_checker.rb]: NP: I would mention a case where a required product volume is missing: `in a UEFI system, or a required product volume (e.g., swap) is missing.`

File: src/lib/y2storage/setup_checker.rb [hunk 5]
```diff
@@ -138,5 +138,9 @@ def mandatory?(volume)
     def missing?(volume)
       BlkDevice.all(devicegraph).none? { |d| d.match_volume?(volume) }
     end
+
+    def boot_requirements_checker
```
Comment (joseivanlopez) [inline src/lib/y2storage/setup_checker.rb]: NP: I know that it is pretty clear, but please, comment at least the return type.

File: src/lib/y2partitioner/setup_errors_presenter.rb [hunk 6]
```diff
@@ -44,37 +44,51 @@ def initialize(setup_checker)
...
+    # HTML representation for mandatory product warnings
     #
-    # @return [String, nil] nil if there is no product error
-    def product_errors_html
-      errors = setup_checker.product_errors
+    # @return [String, nil] nil if there is no product warning
+    def product_warnings_html
+      warnings = setup_checker.product_warnings
+      # TRANSLATORS
+      header = _("The system could not work properly because the following warnings were found:\n")
```
Comment (joseivanlopez) [inline src/lib/y2partitioner/setup_errors_presenter.rb]: NP: I would avoid the word `warnings` there. An option could be: `The system could not work properly because the following issues were found:`

File: src/lib/y2partitioner/setup_errors_presenter.rb [hunk 7]
```diff
@@ -44,37 +44,51 @@ def initialize(setup_checker)
     #
     # @return [String]
     def to_html
```
Comment (joseivanlopez) [inline src/lib/y2partitioner/setup_errors_presenter.rb]: NP: I would simplify the method a little bit by defining a `#warnings_html` in a similar way to `#errors_html`:

```
def to_html
  errors_html || warnings_html || ""
end
```

File: test/y2storage/boot_requirements_errors_test.rb [hunk 8]
```diff
@@ -209,6 +222,9 @@
       allow_any_instance_of(Y2Storage::BootRequirementsStrategies::ZIPL)
         .to receive(:zipl_volume).and_return(zipl_volume)
 
+      allow_any_instance_of(Y2Storage::BootRequirementsStrategies::ZIPL)
+        .to receive(:minimal_zipl_volume).and_return(zipl_volume)
```
Comment (joseivanlopez) [inline test/y2storage/boot_requirements_errors_test.rb]: Is this needed?

File: test/y2storage/boot_requirements_errors_test.rb [hunk 9]
```diff
@@ -268,7 +284,7 @@
         context "when boot device has no partition table" do
           let(:boot_partition_table) { nil }
 
-          it "contains an error for unknown partition table" do
+          it "contains an fatal error for unknown partition table" do
```
Comment (joseivanlopez) [inline test/y2storage/boot_requirements_errors_test.rb]: NP: typo `a fatal`

File: src/lib/y2storage/boot_requirements_strategies/base.rb [hunk 10]
```diff
@@ -68,18 +68,35 @@ def needed_partitions(target)
...
+      #
+      # @note This method can be overloaded for derived classes.
+      #
+      # @see SetupError
+      #
+      # @return [Array<SetupError>]
+      def errors
+        res = []
+        if root_filesystem_missing?
+          error_message = _("There is no device mounted to '/'")
```
Comment (joseivanlopez) [inline src/lib/y2storage/boot_requirements_strategies/base.rb]: NP: `There is no device mounted at '/'` ??

File: src/lib/y2storage/boot_requirements_strategies/legacy.rb [hunk 11]
```diff
@@ -177,8 +183,7 @@ def errors_on_msdos
       def unknown_boot_partition_table_error
         # TRANSLATORS: error message
         error_message = _(
-          "Boot requirements cannot be determined because " \
-          "boot disk has not partition table"
+          "Boot disk is without partition table"
```
[the remark this answers is not in the document]
Comment (jreidinger) [reply src/lib/y2storage/boot_requirements_strategies/legacy.rb]: I do not like too formal "Boot requirements cannot be determined" it sounds like our error that we cannot determine it. I think we should make it more user oriented. How you will see
`Boot disk is without partition table and we does not support booting from it`

BTW that reminds me that it would be nice to also add to end of message how to fix it. E.g. `You can fix it by creating partition table on disk`.

File: src/lib/y2storage/setup_checker.rb (hunk 5 above)
Comment (jreidinger) [reply src/lib/y2storage/setup_checker.rb]: :+1:

File: src/lib/y2partitioner/setup_errors_presenter.rb (hunk 6 above)
Comment (jreidinger) [reply src/lib/y2partitioner/setup_errors_presenter.rb]: makes sense. Maybe we should also mention that it is product specific requirements?

File: test/y2storage/boot_requirements_errors_test.rb (hunk 8 above)
Comment (jreidinger) [reply test/y2storage/boot_requirements_errors_test.rb]: no longer

File: src/lib/y2storage/boot_requirements_strategies/base.rb (hunk 10 above)
Comment (jreidinger) [reply src/lib/y2storage/boot_requirements_strategies/base.rb]: :+1:

File: src/lib/y2partitioner/setup_errors_presenter.rb (hunk 6 above)
Comment (joseivanlopez) [reply src/lib/y2partitioner/setup_errors_presenter.rb]: Yes, it makes sense.

File: src/lib/y2storage/boot_requirements_strategies/legacy.rb [hunk 12]
```diff
@@ -37,23 +37,29 @@ def needed_partitions(target)
...
-      def errors
-        errors = super
-
-        if root_filesystem_missing?
-          errors << unknown_boot_disk_error
-        elsif boot_partition_table_missing?
-          errors << unknown_boot_partition_table_error
-        elsif boot_ptable_type?(:gpt)
-          errors += errors_on_gpt
+      def warnings
```
[the remark this answers is not in the document]
Comment (jreidinger) [reply src/lib/y2storage/boot_requirements_strategies/legacy.rb]: warnings is in fact never called when error is there ( that is reason why original I had exception there )

File: src/lib/y2partitioner/setup_errors_presenter.rb [hunk 13]
```diff
@@ -44,20 +44,24 @@ def initialize(setup_checker)
...
-      errors = errors_html
-      return errors if errors
-
-      warnings = [boot_warnings_html, product_warnings_html].compact
-      return "" if warnings.empty?
-
-      warnings.join(Yast::HTML.Newline)
+      errors_html || warnings_html || nil
```
Comment (joseivanlopez) [inline src/lib/y2partitioner/setup_errors_presenter.rb]: NP: Strictly speaking, last `|| nil` is not needed, but I can live with it.

File: src/lib/y2storage/boot_requirements_strategies/legacy.rb [hunk 14]
```diff
@@ -183,7 +188,8 @@ def errors_on_msdos
       def unknown_boot_partition_table_error
         # TRANSLATORS: error message
         error_message = _(
-          "Boot disk is without partition table"
+          "Boot disk is without partition table and we does not support " \
+            "booting from it. You can fix it by creating partition table on disk."
```
Comment (joseivanlopez) [inline src/lib/y2storage/boot_requirements_strategies/legacy.rb]: NP: I will improve the text a little bit more.

"Boot disk has no partition table and it is not possible to boot from it." \
"You can fix it by creating a partition table on the boot disk."

File: test/y2storage/boot_requirements_errors_test.rb [hunk 15]
```diff
@@ -406,7 +406,18 @@
         context "with a partitions-based proposal" do
           let(:use_lvm) { false }
 
-          include_examples "PReP partition"
```
Comment (joseivanlopez) [inline test/y2storage/boot_requirements_errors_test.rb]: Why is this change needed?
